src/ProductConstructionService/ProductConstructionService.Deployment/Deployer.cs

```diff
@@ -222,33 +237,81 @@ private string GetLogsLink(string revisionName)
 
     private async Task<ProcessExecutionResult> InvokeAzCLI(string[] command)
     {
+        string[] fullCommand = [.. command, .. DefaultAzCliParameters];
+        _logger.LogInformation($"Invoking az cli command `{command}`", string.Join(' ', fullCommand));
         return await _processManager.Execute(
             Path.GetFileName(_options.AzCliPath),
-            [
-                .. command,
-                .. DefaultAzCliParameters
-            ],
+            fullCommand,
             workingDir: Path.GetDirectoryName(_options.AzCliPath));
     }
 
-    private async Task StopProcessingNewJobs()
+    private async Task StopProcessingNewJobs(string activeRevisionName)
     {
-        Console.WriteLine("Stopping the service from processing new jobs");
-        await _pcsClient.Status.StopPcsWorkItemProcessorAsync();
+        _logger.LogInformation("Stopping the service from processing new jobs");
 
-        string status;
+        var replicas = await GetRevisionReplicaStates(activeRevisionName);
         try
         {
+            foreach (var replica in replicas)
+            {
+                await replica.FinishWorkItemAndStopAsync();
+            }
+
+            bool allReplicasStopped;
             do
             {
-                status = await _pcsClient.Status.GetPcsWorkItemProcessorStatusAsync();
+                allReplicasStopped = true;
+                foreach (var replica in replicas)
+                {
+                    var state = await replica.GetStateAsync();
+                    _logger.LogInformation("Replica {replicaName} has status: {state}",
+                        replica.ReplicaName,
+                        state);
+                    if (state != WorkItemProcessorState.Stopped)
+                    {
+                        allReplicasStopped = false;
+                    }
+                }
 
-                Console.WriteLine($"Current status: {status}");
-            } while (await Utility.SleepIfTrue(() => status != "Stopped", SleepTimeSeconds));
+                if (!allReplicasStopped)
+                {
+                    _logger.LogInformation("Waiting for all replicas to stop");
+                }
+            } while (await Utility.SleepIfTrue(() => !allReplicasStopped, SleepTimeSeconds));
```

Comment:
Can this run forever?

Reply:
The code itself, yes, but AzDo has a timeout which kills it. But we can definitely add some limit here too

Reply:
I added a retry limit to it